exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngBatchReader.java

```diff
@@ -0,0 +1,271 @@
+
+  private final PcapngFormatConfig config;
+  private final EasySubScan scan;
+  private final int maxRecords;
+  private CustomErrorContext errorContext;
+  private List<SchemaPath> columns;
+  private List<ColumnDefn> projectedColumns;
+  private Iterator<IPcapngType> pcapIterator;
+  private RowSetLoader loader;
+  private InputStream in;
+  private Path path;
+
+  public PcapngBatchReader(final PcapngFormatConfig config, final EasySubScan scan) {
+    this.config = config;
+    this.scan = scan;
+    this.maxRecords = scan.getMaxRecords();
+    this.columns = scan.getColumns();
+  }
+
+  @Override
+  public boolean open(FileSchemaNegotiator negotiator) {
+    try {
+      // init InputStream for pcap file
+      errorContext = negotiator.parentErrorContext();
+      DrillFileSystem dfs = negotiator.fileSystem();
+      path = dfs.makeQualified(negotiator.split().getPath());
+      in = dfs.openPossiblyCompressedStream(path);
+      // decode the pcap file
+      PcapDecoder decoder = new PcapDecoder(IOUtils.toByteArray(in));
+      decoder.decode();
+      pcapIterator = decoder.getSectionList().iterator();
+      logger.debug("The config is {}, root is {}, columns has {}", config, scan.getSelectionRoot(), columns);
+    } catch (IOException e) {
+      throw UserException
+             .dataReadError(e)
+             .message("Failure in initial pcapng inputstream. " + e.getMessage())
+             .addContext(errorContext)
+             .build(logger);
+    } catch (Exception e) {
+      throw UserException
+             .dataReadError(e)
+             .message("Failed to decode the pcapng file. " + e.getMessage())
+             .addContext(errorContext)
+             .build(logger);
+    }
+    // define the schema
+    negotiator.tableSchema(defineMetadata(), true);
+    ResultSetLoader resultSetLoader = negotiator.build();
+    loader = resultSetLoader.writer();
+    // bind the writer for columns
+    bindColumns(loader);
+    return true;
+  }
+
+  /**
+   * The default of the `stat` parameter is false,
+   * which means that the packet data is parsed and returned,
+   * but if true, will return the statistics data about the each pcapng file only
+   * (consist of the information about collect devices and the summary of the packet data above).
+   *
+   * In addition, a pcapng file contains a single Section Header Block (SHB),
+   * a single Interface Description Block (IDB) and a few Enhanced Packet Blocks (EPB).
+   * <pre>
+   * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   * | SHB | IDB | EPB | EPB |    ...    | EPB |
+   * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   * </pre>
+   * https://pcapng.github.io/pcapng/draft-tuexen-opsawg-pcapng.html#name-physical-file-layout
+   */
+  @Override
+  public boolean next() {
+    while (!loader.isFull()) {
+      if (pcapIterator.hasNext()) {
+        IPcapngType block = pcapIterator.next();
+        if (config.getStat()) {
```

Comment:
@luocooong 
Thanks for addressing the review comments.   I have a few more minor comments and we should be good to go.  This particular logic here is a bit difficult to follow.  Could we refactor the if-statements so that it is a little more clear?

I'm thinking (pseudocode)
```java
if (config.getStat() && isIEnhancedPacketBlock()) {
   // do something
} else if (! config.getStat() && !isEnhancedPacketBlock()) {
  // Do something else
}
```

My personal preference is to put the check to see if the parser has more records at the beginning of the loop.  So it would look something like this: (pseudocode)

```java
while (!loader.isFull() ) {
  // No more records
  if (!parser.hasNext()) {
    return false;
  } else if (config.getStat() && isIEnhancedPacketBlock()) {
     // do something
  } else if (! config.getStat() && !isEnhancedPacketBlock()) {
    // Do something else
  }
}

// process the row
loader.start()
}
```